From 8a71a3e76b1db6412923cb9d0a91bc2de40375cf Mon Sep 17 00:00:00 2001
From: Mohamad Noor Alim Hussin <mnalim@efinixinc.com>
Date: Mon, 18 Dec 2023 14:50:18 +0800
Subject: [PATCH 2/2] mmc.c: add workaround for efinix sd host driver

This workaround for efinix only works for efinix sd
host controller driver.

Signed-off-by: Mohamad Noor Alim Hussin <mnalim@efinixinc.com>
---
 drivers/mmc/mmc.c | 12 ++++++++++--
 1 file changed, 10 insertions(+), 2 deletions(-)

diff --git a/drivers/mmc/mmc.c b/drivers/mmc/mmc.c
index b4c8e7f293..27b157db4a 100644
--- a/drivers/mmc/mmc.c
+++ b/drivers/mmc/mmc.c
@@ -1330,6 +1330,8 @@ static int sd_get_capabilities(struct mmc *mmc)
 		mmc->version = SD_VERSION_2;
 		if ((mmc->scr[0] >> 15) & 0x1)
 			mmc->version = SD_VERSION_3;
+		/* force to use version 2 for efx sd host */
+		mmc->version = SD_VERSION_2;
 		break;
 	default:
 		mmc->version = SD_VERSION_1_0;
@@ -2506,7 +2508,13 @@ static int mmc_startup(struct mmc *mmc)
 	mmc_select_mode(mmc, MMC_LEGACY);
 
 	mmc->dsr_imp = ((cmd.response[1] >> 12) & 0x1);
-	mmc->read_bl_len = 1 << ((cmd.response[1] >> 16) & 0xf);
+	if (IS_SD(mmc)) {
+                if (mmc->high_capacity)
+                        mmc->read_bl_len = 512;
+                else
+                        mmc->read_bl_len = 1 << ((cmd.response[1] >> 16) & 0xf);
+        }
+
 #if CONFIG_IS_ENABLED(MMC_WRITE)
 
 	if (IS_SD(mmc))
@@ -2526,7 +2534,7 @@ static int mmc_startup(struct mmc *mmc)
 	}
 
 	mmc->capacity_user = (csize + 1) << (cmult + 2);
-	mmc->capacity_user *= mmc->read_bl_len;
+	mmc->capacity_user *= 4;
 	mmc->capacity_boot = 0;
 	mmc->capacity_rpmb = 0;
 	for (i = 0; i < 4; i++)
-- 
2.17.1

